proposals/0684-adding-e2e-in-core.md

```diff
@@ -0,0 +1,124 @@
+
+## Motivation
+
+The main "why" has already been explained above: introducing more solid automation to thoroughly verify that code changes don't introduce regressions. This will lead to removing the need for the release crew to do local testing to validate the code before a release.
+
+Another explanation necessary is the stack decision: the reason is that it's the combination that gets the closest to the specific requirements of React Native's unique structure. We needed to fulfil these needs:
+
+- had to be "vanilla Jest" based
+  - in order to very closely match the internal Meta monorepo tooling used by the engineers, and in doing so lower the barrier to entry to add tests for this set too
+- had to not touch RNTester code (pure black box testing)
+- needed to support device farm
+  - while in a first phase the testing will run on CI on emulator/simulators, we are aware of the inherit flakyness so we want to start using a device farm relatively soon during the work on this effort
+- easy to extend to out of tree platforms (ex. Windows, macOS)
+  - RNWindows already uses [a very similar stack](https://github.com/microsoft/react-native-windows/blob/main/docs/e2e-testing.md).
+
+With that specific combination of needs, Appium+WBIO+Jest came out as the only viable option.
+
+## Detailed design
+
+As already mentioned, the core stack involved is composed of [Appium](https://appium.io/) and [WebDriverIO](https://webdriver.io/) and [Jest](https://jestjs.io/).
+
+The dedicated code configuring the tooling and the tests will be in its own private package in `packages/rn-tester-e2e`. The app used for testing is RNTester (in `packages/rn-tester`), an app already present in the codebase which contains code for every component and API with sample implementations.
+
+In it, the `tests` folder is where the tests and referencing files all live. The substructure is as follows:
+
+- `screens` -> in this folder, you will find `*.screen.js` files, where each file represents a navigation screen for RNTester. So there are 3 root ones (`apis`, `bookmarks`, `components`) and then for subscreens, there's a folder with the same name. Each of these files provide an easy way to define all elements present in said screen, so that they can be used for tests.
+  - for each element of a given screen there will be a platform-specific reference in the form of
+
+    ```js
+    btnSubmitElement: Utils.platformSelect({
+      ios: iOSLabel('Press to submit your application!'),
+      android: androidWidget('Button', 'resource-id', 'button_default_styling'),
+    }),
+    ```
+
+    we did, during early investigations, attempt to figure out more streamlined and universal solutions, such as relying on accessibility labels like `testID`s, but it created inconsistent scenarios nor every element on screen always had a `testID` available.
```

Comment:
maybe we can add a note that `iOSLabel` and `androidWidget` are utility functions we created for the test. While writing more tests, we will introduce more utility functions and writing tests will be easier and easier.